Add layer norm out variant

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78586
* __->__ #76149

tugsbayasgalan committed:
Add layer norm out variant

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add layer norm out variant"




[ghstack-poisoned]

vadimkantorov commented:
Related PR: https://github.com/pytorch/pytorch/pull/61821

albanD commented:
Good point @vadimkantorov , Thanks!

@tugsbayasgalan what is the difference between this and https://github.com/pytorch/pytorch/pull/61821 ?

tugsbayasgalan commented:
I wasn't aware of this PR until today. Let me take a look

tugsbayasgalan commented:
> Good point @vadimkantorov , Thanks!
> 
> @tugsbayasgalan what is the difference between this and #61821 ?

It seems similar but the old diff is incomplete because you need to special case for non-contiguous tensors I think

tugsbayasgalan committed:
Update on "Add layer norm out variant"




[ghstack-poisoned]

albanD commented on torch/nn/functional.py:
```diff
@@ -2439,6 +2439,7 @@ def layer_norm(
     weight: Optional[Tensor] = None,
     bias: Optional[Tensor] = None,
     eps: float = 1e-5,
+    out: Optional[Tensor] = None,
```
I'm not sure we want that. I don't think any functional function has out arguments

vadimkantorov replied:
maybe they should? from my experience, there's no big difference between core and functional functions besides historical reasons or just reducing surface of core namespace.

doing out-variant is a useful intermediate step for implementing an inplace autograd function: https://github.com/pytorch/pytorch/issues/26288

albanD replied:
cc @jbschlosser for the more general question here?

jbschlosser replied:
IIRC there is one function in `torch.nn.functional` that takes an `out` param: `normalize()`.

Imo it's not great that the functional wrappers in general restrict users from accessing the out variants. I think ideally either the functional wrappers wouldn't exist, or they would consistently allow for `out` to be specified.

If there's no JIT / other reason preventing us from supporting `out` here, I think we should. Even better - is it possible to remove the `torch.nn.functional.layer_norm` definition and use `_add_docstr()` instead?

ngimel replied:
Agree with @jbschlosser, looks like this is a no-op wrapper, so we shouldn't have it and should have `_add_docstr()`

tugsbayasgalan replied:
sounds good!

jbschlosser replied:
Update: turns out there is a purpose for this wrapper wrt FX. From @jamesr66a:

> yeah the problem here is that tracing functionals is more permissive than tracing ops. When tracing functionals, we can accept a proxy argument for any parameter. when tracing ops, the PythonArgParser will only do _torch_function_ dispatch if proxys appear for parameters that are of Tensor or TensorList type. So by removing the functional and replacing it with an op, this is causing a traceability regression.
I've complained about the issue with tracing ops, but the core team said they do it for performance reasons. Maybe you could try bugging them again about this?

Problem is with the `normalized_size` arg since it's not of type `Tensor` or `TensorList`. If we want to keep FX traceability, I think we have to keep the wrapper here.

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2571,6 +2571,11 @@
+- func: layer_norm.out(Tensor input, int[] normalized_shape, Tensor? weight=None, Tensor? bias=None, float eps=1e-05, bool cudnn_enable=True, *, Tensor(a!) out) -> Tensor(a!)
+  variants: function
+  dispatch:
+    CompositeExplicitAutograd: layer_norm_out
```
out variants don't support autograd. So no need for dispatch key. And you could let the usual c++ overload resolution find your function?

albanD commented on aten/src/ATen/native/layer_norm.cpp:
```diff
@@ -186,6 +189,56 @@ Tensor layer_norm(
+    layer_norm_with_mean_rstd_out(
+        Y, mean, rstd, *X, normalized_shape, *gamma, *beta, eps, M, N);
+    resize_output(output, Y.sizes());
+    output.copy_(std::move(Y));
```
cc @mruberry is that ok?

tugsbayasgalan commented on torch/nn/functional.py:
```diff
@@ -2448,7 +2449,10 @@ def layer_norm(
         return handle_torch_function(
             layer_norm, (input, weight, bias), input, normalized_shape, weight=weight, bias=bias, eps=eps
```
sounds good

tugsbayasgalan commented on aten/src/ATen/native/Copy.cpp:
```diff
@@ -238,6 +238,7 @@ static Tensor & copy_impl(Tensor & self, const Tensor & src, bool non_blocking)
   if(!self.is_complex() && src.is_complex()) {
     TORCH_WARN_ONCE("Casting complex values to real discards the imaginary part");
   }
+
```
oops! will remove

tugsbayasgalan committed:
Update on "Add layer norm out variant"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add layer norm out variant"




[ghstack-poisoned]

vadimkantorov commented:
once out-variant is supported, it would be very cool to also have inplace-variants: https://github.com/pytorch/pytorch/issues/26288

tugsbayasgalan committed:
Update on "Add layer norm out variant"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add layer norm out variant"




[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36216336).

tugsbayasgalan committed:
Update on "Add layer norm out variant"


Differential Revision: [D36216336](https://our.internmc.facebook.com/intern/diff/D36216336)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add layer norm out variant"


Differential Revision: [D36216336](https://our.internmc.facebook.com/intern/diff/D36216336)

[ghstack-poisoned]

jbschlosser commented on aten/src/ATen/native/layer_norm.cpp:
```diff
@@ -186,6 +189,56 @@ Tensor layer_norm(
+        c10::nullopt /* device */,
+        c10::nullopt /* pin_memory */,
+        at::MemoryFormat::Contiguous);
+    layer_norm_with_mean_rstd_out(
```
For this to work with CUDA, you should dispatch to `LayerNormKernel` instead of `layer_norm_with_mean_rstd_out`, as the latter hard-codes a call to `LayerNormKernel` for CPU:
https://github.com/pytorch/pytorch/blob/ec33f0ee838e5b36089afea9d230e46f205a0a49/aten/src/ATen/native/layer_norm.cpp#L21-L36

tugsbayasgalan committed:
Update on "Add layer norm out variant"


Differential Revision: [D36216336](https://our.internmc.facebook.com/intern/diff/D36216336)

[ghstack-poisoned]

jbschlosser commented on aten/src/ATen/native/layer_norm.cpp:
```diff
@@ -186,6 +189,71 @@ Tensor layer_norm(
+        N,
+        eps,
+        &Y,
+        /*mean=*/nullptr,
+        /*rstd=*/nullptr);
```
Even if you don't need to compute these (since out variants don't support autograd), the CUDA kernel still assumes that something is passed for them:
https://github.com/pytorch/pytorch/blob/bbb1a55d9f254ad5f587af290b2c975e6f505af9/aten/src/ATen/native/cuda/layer_norm_kernel.cu#L680-L681

I think this explains the segfaults you're getting in the failing CI checks. Note that the out-of-place CPU kernel does this that you can probably copy:
```cpp
Tensor mean = at::empty({M}, X->options());
Tensor rstd = at::empty({M}, X->options());
```

tugsbayasgalan replied:
I see thanks :)

tugsbayasgalan committed:
Update on "Add layer norm out variant"


Differential Revision: [D36216336](https://our.internmc.facebook.com/intern/diff/D36216336)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add layer norm out variant"



[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add layer norm out variant"



[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add layer norm out variant"



[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add layer norm out variant"



[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add layer norm out variant"



[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add layer norm out variant"



[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add layer norm out variant"



[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add layer norm out variant"



[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add layer norm out variant"



[ghstack-poisoned]

jbschlosser reviewed (commented):
Looks good now, but I'll still suggest going the proper way by using structured kernels
